CP15 Cache Emulation Work

DesperateProgrammer opened the pull request:
The instruction and data cache register were not yet implemented in MelonDS and would act different against the hardware by just reading/writing them.

They have no impact on the - not yet implemented - data cache nor on the instruction cache until their implementation progresses.

DesperateProgrammer committed:
Added CP15 Data and Instruction Cache Lockdown Register

DesperateProgrammer committed:
Merge branch 'melonDS-emu:master' into mem9_timings

JesseTG commented on src/ARM.h:
```diff
@@ -322,6 +322,7 @@ class ARMv5 : public ARM
     u32 RNGSeed;
 
     u32 DTCMSetting, ITCMSetting;
+    u32 DCacheLockDown, ICacheLockDown;
```
Don't forget to add these to the savestate (and to increment the savestate version).

DesperateProgrammer committed:
Added ICacheLockDown and DCacheLockDown to CP15 savestate

JesseTG commented on src/CP15.cpp:
Thank you. Could you please increment the major version number in `Savestate.h` as well? It's our only way to detect savestate format changes at runtime; if it's not updated, then loading old states may crash (or worse, silently break the emulated console).

DesperateProgrammer committed:
Updated savestate version

DesperateProgrammer commented:
I'll add some more changes to the CP15 Caching Instructions and Handling here.

DesperateProgrammer committed:
Cleaned up magic numbers and simplified (not yet used) ICache functions
Marked reading CP15 Cache Dirty Bit as not present

DesperateProgrammer committed:
Enable instruction cache routines. Fixed typos in constants.

DesperateProgrammer committed:
Replaced more CP15 magic values with named constants

DesperateProgrammer committed:
Implemented "weird" instruction cache invalidation by Set/Way

DesperateProgrammer commented:
The last commit was a bit bigger / combined, as adding DCache required other things to not break some applications.

Instruction cache is working except:
 - Prefetch via CP15 (todo)
 - CacheLockdown has no effect yet
 - Timings on cache-fill are not correct yet
Data cache is working except:
- Timings on cache fill are not correct yet
- Timings on Cache hit are not correct yet
- CacheLockdown has no effect yet
- WriteBack is not implemented yet (it writes through and invalidates the cache line atm)

CP15 Registers Operations:
- Cache Clean has no effect yet (needs write back mechanism)
- Cache Clean & Invalidate will now also invalidate the cache (was missing)
- Cache Debug Index Register was added
- ICache/DCache Tag Registers were added
- ICache/DCache Data Registers are not done yet
- Privileged/User Checks are missing

DesperateProgrammer committed:
Implemented DCache and several CP15 registers

DesperateProgrammer committed:
Added Cache Data Write and Read via CP15

DesperateProgrammer committed:
Fixed data cache using only 1 cycle on miss.

DesperateProgrammer committed:
Changed write to cached data from invalidating to updating the cached data to get nearer to hw timings

DesperateProgrammer committed:
Fixed unaligned access to data cache

DesperateProgrammer committed:
Disabled Caches, when JIT is enabled

DesperateProgrammer committed:
Implemented CacheLockDown

DesperateProgrammer commented:
Update on the status

Instruction cache is working except:
-     Prefetch via CP15 (todo)
-     Timings on cache-fill are not correct yet

Data cache is working except:
-     Timings on cache fill are not correct yet
-     WriteBack is not implemented yet (it writes through)

CP15 Registers Operations:
-     Cache Clean has no effect yet (needs write back mechanism)
-     Cache Clean & Invalidate will now also invalidate the cache (was missing)
-     Cache Debug Index Register is working
-     ICache/DCache Tag Registers are working
-     ICache/DCache Data Registers are working
-     Privileged/User Checks are missing

JIT:
- Caches are now disabled when JIT is enabled and no longer breaks JIT.

DesperateProgrammer committed:
Merge branch 'melonDS-emu:master' into mem9_timings

DesperateProgrammer committed:
Added CP15 prefetch routine

DesperateProgrammer committed:
Removed magic number from Cache Fill timings.

DesperateProgrammer committed:
Added permission checks to CP15 Cache register write operations

DesperateProgrammer committed:
Added privilege checks for reading & writing CP15 cache registers

DesperateProgrammer committed:
Replaced Magic numbers in CP15 cache type register

DesperateProgrammer committed:
Fixed Typo and missing file on the cache type constants

DesperateProgrammer committed:
Changed DCache Random Cache-Line selection to a double Galoise LFSR

DesperateProgrammer committed:
Simplified set selection adjustment for the cache lock down

DesperateProgrammer committed:
Cleaned up and fastened up

DesperateProgrammer committed:
Further clean up, removal of magic numbers from CP15.cpp
Split the cp15 constants into CP15_Constants.h instead from MemConstants.

DesperateProgrammer committed:
Removed unneccessary wfcsettings file

DesperateProgrammer committed:
Added CP15 Trace Process ID

DesperateProgrammer committed:
Added check of op1 in MCR/MRC
Fixed Cache Debug registers were accessible, when op1 != 3 in MCR/MRC instructions
Added BIST Test Status register and its cache linefill disable bits

DesperateProgrammer committed:
Included the I/DCache Streaming disable bits in cache lookup

DesperateProgrammer committed:
Added several doxygen-style comments for documentation

DesperateProgrammer committed:
Removed Thumb Check on CP15 Access restriction as MCR/MRC are not present in thumb

DesperateProgrammer committed:
Cleaned up some more magic numbers
Fixed a bug causing overlapping protection regions priority not taken into account, when access permission or cachability bits were changed only on the least priority overlap

DesperateProgrammer committed:
Added more function documenting comments
Added const properties to the CP15Write/Read functions

DesperateProgrammer committed:
Added more documenting comments

DesperateProgrammer committed:
Added more documenting comments
Replaced mogic values with named constants
Added const specifier to some argument and subsequent calls

DesperateProgrammer committed:
Merge branch 'melonDS-emu:master' into mem9_timings

DesperateProgrammer committed:
Added write-back ability in addition to write-through for the data cache

DesperateProgrammer committed:
Removed some debug remains
Updated documenting comments for the DCacheClear* methods

DesperateProgrammer committed:
Fixed an issue with caclulating DTCM/ITCM masks after addr was declared constant
Encapsuled the cache features in #if to disable the features via compile flags

DesperateProgrammer committed:
Enabled Overlapping region fix again

DesperateProgrammer commented:
This turned out to be quite some rework for the CP15.
The Cache is now working including the write-back mode. Cache Debug registers, BIST Cache Test register
CP15 Registers and Functions received some documenting comments.
Many CP15 related calls received the const-specifier for its arguments when they are not changed in the call.

RSDuck commented on src/CP15.cpp:
```diff
@@ -332,153 +366,596 @@ u32 ARMv5::RandomLineIndex()
 
-    id <<= 2;
-    if (ICacheTags[id+0] == tag)
+    for (int set=0;set<ICACHE_SETS;set++)
```
can you add spaces after the semicolons in all the for loops?

RSDuck commented on src/CP15.cpp:
```diff
@@ -332,153 +366,596 @@ u32 ARMv5::RandomLineIndex()
+    // Do the same as above but instead of using if-else 
+    // utilize the && and || operators to skip parts of the operations
+    // With the order of comparison we can put the most likely path
+    // checked first
```
I'm not really sure whether this abusing shortcircuit evaluation really gives better code. If you want to steer the compiler using likely and unlikely markers could be used for the ifs.

DesperateProgrammer replied:
I have removed this and added the [[(un)likely]] attributes

RSDuck commented on src/CP15.cpp:
```diff
@@ -332,153 +366,596 @@ u32 ARMv5::RandomLineIndex()
+    for (int i = 0; i < DCACHE_LINELENGTH; i+=sizeof(u32))
+    {
+        if (tag+i < ITCMSize)
+        {
```
oh so you can actually put ITCM and DTCM into cache?

DesperateProgrammer replied:
I am not entirely sure (I think ) I had a block diagram showing the access passing the write buffer before the TCMs but can't seem to find that diagram. If that is the case, the TCM access will go through the cache when enabled.
However my decision to place the read from TCMs here was to keep the access flow within the line fills in sync to the DataReadXX functions.

RSDuck commented on src/CP15.cpp:
```diff
@@ -332,153 +366,596 @@ u32 ARMv5::RandomLineIndex()
+                {
+                    if (addr & (DCACHE_LINELENGTH / 2))
+                    {
+                        DCacheTags[id+set] |= CACHE_FLAG_DIRTY_UPPERHALF ;
```
please remove the spaces before semicolon

RSDuck commented on src/CP15.cpp:
```diff
@@ -306,7 +340,7 @@ void ARMv5::UpdateRegionTimings(u32 addrstart, u32 addrend)
-        if (pu & 0x10)
+        if (pu & CP15_MAP_DCACHEABLE)
         {
             MemTimings[i][1] = kDataCacheTiming;
```
these values will only ever be used when JIT is enabled, right?

DesperateProgrammer replied:
Right now, yes. They are used in ARMv5::CodeRead32, when the JIT is not enabled.

DesperateProgrammer committed:
Formatting corrections
Removed premature optimization and replaced them with [[(un)likely]]

DesperateProgrammer committed:
Merge branch 'melonDS-emu:master' into mem9_timings

KostaSaizo7 commented:
Fixes https://github.com/melonDS-emu/melonDS/issues/380

KostaSaizo7 commented:
Unfortunately https://github.com/melonDS-emu/melonDS/issues/359 is still occuring

Jaklyy commented:
seems like disabling the pu via the control reg causes melonds to hang under this pr?
at least reproduceable under this test rom:
[zboxtesttest.zip](https://github.com/melonDS-emu/melonDS/files/15212734/zboxtesttest.zip)

DesperateProgrammer commented:
Thanks for the report and sample Jakly,

I'll check that as soon i got some time.

Jaklyy commented on src/CP15.cpp:
```diff
@@ -332,153 +366,557 @@ u32 ARMv5::RandomLineIndex()
-    CodeCycles = (NDS.ARM9MemTimings[addr >> 14][2] + (NDS.ARM9MemTimings[addr >> 14][3] * 7)) << NDS.ARM9ClockShift;
-    CurICacheLine = ptr;
+    //                      first N32                                  remaining S32
+    CodeCycles = (NDS.ARM9MemTimings[tag >> 14][2] + (NDS.ARM9MemTimings[tag >> 14][3] * ((DCACHE_LINELENGTH / 4) - 1))) << NDS.ARM9ClockShift;
```
minor error: DCACHE_LINELENGTH should be ICACHE_LINELENGTH
